tensorflow/lite/kernels/activations.cc

```diff
@@ -363,23 +365,30 @@ TfLiteStatus LeakyReluPrepare(TfLiteContext* context, TfLiteNode* node) {
 
   LeakyReluOpData* data = reinterpret_cast<LeakyReluOpData*>(node->user_data);
 
-  if (output->type == kTfLiteUInt8) {
-    const auto* params =
-        reinterpret_cast<TfLiteLeakyReluParams*>(node->builtin_data);
-    // Quantize the alpha with same zero-point and scale as of input
-    data->q_alpha = static_cast<uint8_t>(std::max<float>(
-        std::numeric_limits<uint8_t>::min(),
-        std::min<float>(std::numeric_limits<uint8_t>::max(),
-                        std::round(input->params.zero_point +
-                                   (params->alpha / input->params.scale)))));
-
-    double real_multiplier =
-        input->params.scale * input->params.scale / output->params.scale;
-    QuantizeMultiplierSmallerThanOneExp(
-        real_multiplier, &data->output_multiplier, &data->output_shift);
-  }
-  return context->ResizeTensor(context, output,
-                               TfLiteIntArrayCopy(input->dims));
+  if (output->type == kTfLiteUInt8 || output->type == kTfLiteInt8) {
+    const auto* params = reinterpret_cast<TfLiteLeakyReluParams*>(node->builtin_data);
+    // Quantize the alpha with predetermined ZOOM_FACTOR and ZERO_POINT.
+    // Since in most cases 0 < alpha < 1, by setting ZOOM_FACTOR to be 200,
+    // quantized alpha can fit in INT8 range, and preserve most precision.
+    TF_LITE_ENSURE(context, params->alpha >= 0);
+    TF_LITE_ENSURE(context, params->alpha < 1);
+    static const uint8_t ZOOM_FACTOR = 200;
```

Comment:
why you need to change this implementation?

what's the reason to hard code zero point & scale?

Reply:
Hi @renjie-liu ,
This piece of code is used to quantize alpha. Originally the scale of alpha was chosen to be the same as input scale (`input->params.scale`). However, the input scale can vary a lot depending on the the distribution of the input. This can cause a lot of potential problems. For example, if the input scale is close to 1, and alpha_float = 0.1, then the quantization of alpha will truncate q_alpha to zero_point, which will introduce huge error.

Therefore, I think it is more appropriate to choose the scale of alpha independent of input scale. The value of 200 was chosen because it is the largest within UINT8 limit and divisible by 100, since mostly alpha_float was chosen to be 0.1, 0.01, 0.05, etc. Choosing a number that is divisible by 100 minimize the loss of quantization.

Reply:
why input_scale close to 1 and alpha close to 0.1 will introduce huge error?

output_float = input_float > 0 ? input_float : input_float * alpha

input_float = (input_q - input_zp) * input_scale

output_float = (output_q - output_zp) * output_scale

=>
if input_q - input_zp > 0:
(output_q - output_zp) * output_scale = (input_q - input_zp) * input_scale => 
output_q = (input_q -input_zp) * (input_scale / output_scale) + output_zp

if input_q - input_zp <= 0:
(output_q - output_zp) * output_scale = (input_q - input_zp) * input_scale * alpha =>
output_q = (input_q -input_zp) * (input_scale / output_scale * alpha) + output_zp

Looks to me we only need an additional set of output_multiplier & shift (we don't need q_alpha).

Does that makes sense?

Reply:
Hi @renjie-liu 

That makes sense... I pushed a new commit for your review. Thank you!